platform/openide.util.ui/src/org/openide/util/Utilities.java

```diff
@@ -1357,17 +1357,50 @@ private static Frame findMainWindow()
      */
     // PR4739
     public static Component findDialogParent() {
-        Component parent = KeyboardFocusManager.getCurrentKeyboardFocusManager().getFocusOwner();
+        return findDialogParent(null);
+    }
+
+    /**
+     * Finds an appropriate component to use for a dialog's parent. Similar to {@link #findDialogParent()}
+     * with the ability to specify a suggested parent component.
+     *
+     * @param suggestedParent the component to return if suitable
+     * @return A suitable parent component for swing dialog displayers.
+     *
+     * @see #findMainWindow()
```

Comment:
`findMainWindow()` is private, probably shouldn't be referenced in javadoc. Could just say the main window is the default if nothing better is found.

Reply:
I can remove this reference. It may also make sense to expose it publicly and call it from `ViewHierarchy.getMainWindow()`. However, that brings up another thought that if there is an alternative implementation registered for `WindowManager` that implements `getMainWindow()` differently, using this `Utilities` version in `DialogDisplayerImpl` could cause a regression. As such, it may be better to have a default provider or have this method return `Optional<Component>`. That way `DialogDisplayerImpl` can provide the appropriate `getMainWindow()` implementation as a fallback.

```java
public static Component findDialogParent() {
    return findDialogParent(null, Utilities::findMainWindow);
}

public static Component findDialogParent(Component suggestedParent, Supplier<Component> fallback) {
    Component parent = suggestedParent;
    if (parent == null) {
        parent = KeyboardFocusManager.getCurrentKeyboardFocusManager().getFocusOwner();
    }
    Window active = KeyboardFocusManager.getCurrentKeyboardFocusManager().getActiveWindow();
    if (parent == null) {
        parent = active;
    } else if (active instanceof Dialog && ((Dialog) active).isModal()) {
        Window suggested = parent instanceof Window ? (Window) parent : SwingUtilities.windowForComponent(parent);
        if (suggested != active) {
            return active;
        }
    }
    if (parent == null) {
        return fallback.get();
    }
    return parent;
}
```

Then `DialogDisplayerImpl` can provide the fallback to the main window via `Utilities.findDialogParent(null, WindowManager.getDefault()::getMainWindow)`.

Reply:
> if there is an alternative implementation registered for `WindowManager` that implements `getMainWindow()` differently, using this `Utilities` version in `DialogDisplayerImpl` could cause a regression

I don't see that as an alternative implementation; the difference between *get* and *find* is significant. `findMainWindow()` under no circumstance creates an `NbMainWindow` if one does not exist; it's a hack for some low level parenting decisions that I believe should remain hidden. This method can be used during startup, before there is a main window or even a `WindowManager` (I think).

Reply:
> it may be better to have a default provider or have this method return `Optional<Component>`. That way `DialogDisplayerImpl` can provide the appropriate `getMainWindow()` implementation as a fallback.

I don't see how fallback adds much. At the call site something like
    if((parent = Utilities.findDialogParent(null)) == null)
        parent = fallback;
is equivalent (unless you believe that there is both a main window not named `NbMainWindow` and that there is frame with the name `NbMainWindow` which is not the main window). I wonder if this is the only spot where an alternative implementation of `WindowManager` might present an issue, and even if it is, is this the best way to deal with it? Seems like putting a lot of work into a method that isn't used anywhere and should be part of a broader strategy. That said, if the fallback goes in, I'd suggest handling `fallback == null` to *find* "NbMainWindow".

Reply:
> That said, if the fallback goes in, I'd suggest handling `fallback == null` to _find_ "NbMainWindow"

Doing that would silently ignore the fallback supplier. I'd rather have it throw in this case so you know you didn't specify a valid supplier or can use `findDialogParent()` instead.

Reply:
> I wonder if this is the only spot where an alternative implementation of `WindowManager` might present an issue, and even if it is, is this the best way to deal with it? Seems like putting a lot of work into a method that isn't used anywhere and should be part of a broader strategy.

> noticed that in NbPresenter.java there's findFocusedWindow() which falls back to "NbMainWindow"

This should be considered as part of a general solution. I think the usage/meaning of `findDialogParent(Component suggestedParent, Supplier<Component> fallback)` in the context of alternate window managers needs to be addressed before this becomes a public API.

Possible issue, Utilities can not depend on core.windows as discussed in https://github.com/apache/netbeans/pull/5280#discussion_r1080908550; passing in a method from a module that this module does not depend on would probably get a runtime error.